pydantic/json.py

```diff
@@ -98,10 +98,13 @@ def pydantic_encoder(obj: Any) -> Any:
 def custom_pydantic_encoder(type_encoders: Dict[Any, Callable[[Type[Any]], Any]], obj: Any) -> Any:
     # Check the class type and its superclasses for a matching encoder
     for base in obj.__class__.__mro__[:-1]:
-        try:
+        if base in type_encoders:
             encoder = type_encoders[base]
-        except KeyError:
+        elif base.__name__ in type_encoders:
+            encoder = type_encoders[base.__name__]
+        else:
             continue
```

Comment:
To make this as performant as possible, I think we should do something like

```py
try:
    encoder = type_encoders[base]
except KeyError:
    try:
        encoder = type_encoders[base.__name__]
    except KeyError:
        continue
```

This avoids duplicated lookups.